systemservices/resolver/resolver.go

```diff
@@ -0,0 +1,66 @@
+package resolver
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/mesg-foundation/core/api"
+)
+
+// Resolvers tasks.
+const (
+	addPeersTask string = "AddPeers"
+	resolveTask         = "Resolve"
+)
+
+// Resolver is the system service responsible for getting the address of
+// other node of the network that are running the desired service.
+type Resolver struct {
+	api       *api.API
+	serviceID string
+}
+
+// New creates a new instance of the Resolver system service.
+func New(serviceID string, api *api.API) *Resolver {
+	return &Resolver{
+		api:       api,
+		serviceID: serviceID,
+	}
+}
+
+// AddPeers is the task that actually new add peers to the service.
+func (r *Resolver) AddPeers(addresses []string) error {
+	e, err := r.api.ExecuteAndListen(r.serviceID, addPeersTask, map[string]interface{}{
+		"addresses": addresses,
+	})
+	if err != nil {
+		return err
+	}
+
+	switch e.Output {
+	case "success":
+		return nil
+	case "error":
+		return errors.New(e.OutputData["error"].(string))
+	default:
+		return fmt.Errorf("unexpected output %s", e.Output)
+	}
+}
+
+// Resolve is the task that return the address of a core that runs the desired service.
+func (r *Resolver) Resolve(serviceID string) (address string, err error) {
+	e, err := r.api.ExecuteAndListen(r.serviceID, resolveTask, map[string]interface{}{
+		"serviceID": serviceID,
+	})
+
+	switch e.Output {
+	case "found":
+		return e.OutputData["address"].(string), nil
+	case "notFound":
+		return "", fmt.Errorf("address for service id %s not found", serviceID)
+	case "error":
+		return "", errors.New(e.OutputData["error"].(string))
```

Comment:
What about creating one error type that can contains the error message but also the `outputKey`?
So we don't have a 1000s different error type but we can access the `outputKey` returned by the task.

Reply:
For this one, I don't quite understand what do you mean? The line you highlighted is about _error_ output key and it's only returned from the service when there is an internal error. Output key is always the same but error message can be different depending on the internal error happened in the service.

But I think we need a custom error type for _notFound_ output and [here](https://github.com/mesg-foundation/core/pull/535/commits/6b1986d8e808fd58b5bc0c554e3afd1aa41dbf0f) is the implementation.